subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -335,6 +338,76 @@ public void stopWatchingAfterError() {
         }
     }
 
+    private static class FilterChangesToOutputsChangesHandler implements FileWatcherRegistry.ChangeHandler {
+        private final LocationsWrittenByCurrentBuild locationsWrittenByCurrentBuild;
+        private final FileWatcherRegistry.ChangeHandler delegate;
+
+        public FilterChangesToOutputsChangesHandler(LocationsWrittenByCurrentBuild locationsWrittenByCurrentBuild, FileWatcherRegistry.ChangeHandler delegate) {
+            this.locationsWrittenByCurrentBuild = locationsWrittenByCurrentBuild;
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void handleChange(FileWatcherRegistry.Type type, Path path) {
+            if (!locationsWrittenByCurrentBuild.wasLocationWritten(path.toString())) {
+                delegate.handleChange(type, path);
+            }
+        }
+
+        @Override
+        public void stopWatchingAfterError() {
+            delegate.stopWatchingAfterError();
+        }
+    }
+
+    private class InvalidateVfsChangeHandler implements FileWatcherRegistry.ChangeHandler {
+        @Override
+        public void handleChange(FileWatcherRegistry.Type type, Path path) {
+            rootReference.update(root -> updateNotifyingListeners(
+                diffListener -> root.invalidate(path.toString(), new VfsChangeLoggingNodeDiffListener(type, path, diffListener))
+            ));
+        }
+
+        @Override
+        public void stopWatchingAfterError() {
+            stopWatchingAndInvalidateHierarchyAfterError();
+        }
+    }
+
+    private class BroadcastingChangeHandler implements FileWatcherRegistry.ChangeHandler {
+        @Override
+        public void handleChange(FileWatcherRegistry.Type type, Path path) {
+            if (changeBroadcaster != null) {
+                changeBroadcaster.handleChange(type, path);
+            }
+        }
+
+        @Override
+        public void stopWatchingAfterError() {
+            if (changeBroadcaster != null) {
+                changeBroadcaster.stopWatchingAfterError();
+            }
+        }
+    }
+
+    private static class CompositeChangeHandler implements FileWatcherRegistry.ChangeHandler {
+        private final List<FileWatcherRegistry.ChangeHandler> handlers;
+
+        public CompositeChangeHandler(List<FileWatcherRegistry.ChangeHandler> handlers) {
+            this.handlers = handlers;
+        }
```

Comment:
Nit: I think this should take varargs for nicer call semantics. That way we'd be creating the `ImmutableList` here, which is better for encapsulation. With the current code it is possible to pass in a mutable `List` for example.

Reply:
Done.